course02/project01/gavin/Asteroids/game_world.cc

```diff
@@ -0,0 +1,52 @@
+#include "game_world.h"
+#include "asteroid.h"
+#include "ship.h"
+#include <glm/gtc/random.hpp>
+
+using namespace std::chrono_literals;
+
+GameWorld::GameWorld(Controller &controller) :
+        GameObject(),
+        mController(controller){
+    CreateGameObject<Ship>(ci::app::getWindowCenter(), glm::vec2{0.0, 1.0}, mController);
+    RegisterCallback([this](){
+        SpawnAsteroid();
+    }, 5s);
+}
+
+void GameWorld::Update(const float frameDelta)
+{
+    UpdateCollisions();
+    GameObject::Update(frameDelta);
+}
+
+void GameWorld::Draw()
+{
+    GameObject::Draw();
+}
+
+void GameWorld::SpawnAsteroid()
+{
+    auto direction = glm::sphericalRand(1.0);
+    CreateGameObject<Asteroid>(glm::vec2{1.0, 1.0}, glm::normalize(direction), 20, 30);
+    RegisterCallback([this](){
+        SpawnAsteroid();
+    }, 10s);
+}
+
+void GameWorld::AddCollider(Collidable* collidable)
+{
+    mColliders.push_back(collidable);
+}
+
+void GameWorld::UpdateCollisions() {
+    //FIXME: collection of references without ownership?
+    for(auto a : mColliders)
+        for(auto b : mColliders)
+            if(a != b and a->Overlaps(*b))
+                a->Collide(*b);
+}
```

Comment:
This is dodgy for two reasons - I call a 'Collide' function which could destroy the contents of mColliders while I iterate over it. I also use raw pointers here (I don't want ownership) and just assume that all colliders remove themselves from this vector when they are destroyed.

Reply:
Fixed in 76db252356e796ace0396de1cb5a68c9790e3d06. I use the ```ForAllOf<Collidable>``` method on the elements directly. Collidables no longer need to handle their own registration on the world.